### PR TITLE
Fail early when connection with bastion host fails

### PR DESCRIPTION
**Pull request by sajith**
Issue is #151.  

Currently `Slice.submit()` (and possibly `Slice.modify()` too) will spin on `wait_ssh()`, with repeated attempts every 20 seconds for for 1800 seconds.  With this change, when connection with bastion host fails, we fail early and report the failure.

### CODE REVIEW

File: fabrictestbed_extensions/fablib/slice.py [hunk 1]
```diff
@@ -1506,6 +1514,69 @@ def wait_ssh(self, timeout: int = 1800, interval: int = 20, progress: bool = Fal
...
+        except paramiko.SSHException as e:
+            # Unsure how to handle this. Same as above maybe?
+            logging.error(f"Bastion SSH error: {e}")
+            raise e
+
+        except Exception as e:
+            # Could this be a transient error? Should we keep
+            # re-trying in that case?
+            logging.error(f"Bastion connection error: {e}")
+            raise e
```
Comment (sajith) [inline fabrictestbed_extensions/fablib/slice.py:1575]: I am not sure what the right thing to do with different types of exceptions here.  Should we treat all failures the same, and give up as soon as there is an error?  Or should we attempt connecting again in some cases?

**Comment by sajith** [comment]
Did some testing (with `python tests/integration/test_hello_fabric.py`), and this works as it should:

- When I use the correct bastion key in `~/work/fabric_config/fabric_rc`, the test worked as expected.
- When I use the wrong bastion key , `Slice.submit()`, the test fails early rather than spinning for 30 minutes.

I am a little unclear on how to propagate the failure to talk to bastion (is that an error that can be logged and then ignored, or should we re-throw an error?), and when we should give up and when to re-try.

File: fabrictestbed_extensions/fablib/slice.py (hunk 1 above)
Comment (sajith) [reply fabrictestbed_extensions/fablib/slice.py]: Moved this method to fablib, after discussion with Paul.

**Review by ibaldin — COMMENTED** [review]
Is there a small test harness for it? - this is hard to do automatically, I know.

**Comment by sajith** [comment]
> Is there a small test harness for it? - this is hard to do automatically, I know.

Good catch, thank you.  I will try to figure something out.  It can't be a unit test because FablibManager constructor needs the network, and refactoring that would be out of this PR's scope.  It will have to be an integration test.  

I did some tests with [tests/integration/test_hello_fabric.py](https://github.com/fabric-testbed/fabrictestbed-extensions/blob/main/tests/integration/test_hello_fabric.py) (with one good key, and a set of wrong keys).  Perhaps there can be a stand-alone test.

Sphinx setup also needs to be told of the new public method.

**Comment by sajith** [comment]
> Should I be doing changelog updates too on my PRs

Well, if you want to!  Updating changelog in the PR helps to avoid having to update it at the time of release, but if you find that it isn't worth the trouble, feel free to skip that step.

There is a CI check to remind us, and tagging the PR with `no changelog` at the PR is created (not later because of a GitHub quirk) should omit that check.

**Pull request merged**